diagnostics.js

```diff
@@ -0,0 +1,213 @@
+/**
+ * Contains all configured adapters for the given environment.
+ *
+ * @type {Array}
+ * @public
+ */
+var adapters = [];
+
+/**
+ * Contains all modifier functions.
+ *
+ * @typs {Array}
+ * @public
+ */
+var modifiers = [];
+
+/**
+ * Our default logger.
+ *
+ * @public
+ */
+var logger = function devnull() {};
+
+/**
+ * Register a new adapter that will used to find environments.
+ *
+ * @param {Function} adapter A function that will return the possible env.
+ * @returns {Boolean} Indication of a successful add.
+ * @public
+ */
+function use(adapter) {
+  if (~adapters.indexOf(adapter)) return false;
+
+  adapters.push(adapter);
+  return true;
+}
+
+/**
+ * Assign a new log method.
+ *
+ * @param {Function} custom The log method.
+ * @public
+ */
+function set(custom) {
+  logger = custom;
+}
+
+/**
+ * Check if the namespace is allowed by any of our adapters.
+ *
+ * @param {String} namespace The namespace that needs to be enabled
+ * @returns {Boolean|Promise} Indication if the namespace is enabled by our adapters.
+ * @public
+ */
+function enabled(namespace) {
+  var async = [];
+
+  for (var i = 0; i < adapters.length; i++) {
+    if (Object.prototype.toString.call(adapters[i]) === '[object AsyncFunction]') {
```

Comment:
thought: wouldn't it be easier to just always force adapters to expose a `async` function? Removes the requirement for this dance.

Reply:
The reason why **async** adapters are considered an afterthought is that it adds so much more complexity and weird edge cases to the project. The current flow right now for sync adapters is:

```
debugger is created -> check if allows -> return yep() or nope()
```

With an async system, we can't just return a `yep` or `nope` because we have to wait for the adapters to be resolved. That would also mean that any `debug()` call right after it's been initialized will always be lost on an async system, that's why we have a queue for the RN version so it gets processed once the adapter is resolved. But that also means we cannot return an empty function, so the debug statement can't get optimized away by the JavaScript runtime so you will have a performance hit. We also can't make debugger itself `async` because you can't:

```
const diagnostics = require('diagnostics');
const debug = await diagnostics('namespace');
```

As the module environment is not async/await compatible. Right now, React-Native is the only environment that requires an async adapter because `AsyncStorage` is really the only decent way to retrieve the debug flags.